docs/proposals/ssh_adapter.md

```diff
@@ -0,0 +1,194 @@
+# SSH protocol proposal
+
+We'd like to implement a protocol for Git LFS that uses SSH protocol
+exclusively, avoiding the need to use HTTPS altogether.  This will make
+deployment and use easier in a variety of situations where access to certain
+ports is limited.
+
+## What not to do
+
+There are several possible approaches that could be adopted.  SSH provides a
+native capability for the SFTP protocol, which can be used to transfer files.
+However, in order to implement this on the server side, each access (upload or
+download) must have an access control check instead of one at the beginning of
+the operation.  This would be inefficient in some server-side implementations,
+and nearly impossible to implement securely for implementations that use the
+system OpenSSH for implementation.
+
+## A more usable approach
+
+Git already has some places we can look for inspiration.  Its SSH protocol is
+based on the Git native protocol, which is based on the pkt-line scheme.
+Recently, Git has learned about protocol version 2, which provides better
+support for expressing and negotiating capabilities.
+
+Ideally, we would allow multiple operations to occur on a single connection for
+efficiency's sake, especially on high-latency connections, where the cost of SSH
+connection setup may be high due to multiple round trips.  In addition, a
+protocol which maps well onto HTTP may be beneficial for those server-side
+implementations which would like to proxy connections to an HTTP-based backend.
+
+This design assumes a reference to Git's pkt-line and protocol v2 documentation.
+pkt-line headers for this document may contain values up to 65519 decimal.
+
+To initiate a connection, Git LFS should run the folllowing command:
+
+    $ ssh [{user}@]{server} git-lfs-transfer {path} {operation}
+
+If authentication fails, or some other connection error occurs, errors will be
+read fronm standard error and displayed to the user.  The operation may be
+`upload`, `download`, or `meta` (for items which are not transfers).
+
+Once the connection is established, the server should send a capability
+advertisement:
+
+```
+capability-advertisement = capability-list flush-pkt
+capability-list = *capability
+capability = PKT-LINE(key=[value] LF)
+
+key = 1*(ALPHA | DIGIT | "-_")
+value = 1*(ALPHA | DIGIT | " -_.,?\/{}[]()<>!@#$%^&*+=:;")
+```
+
+Unlike the Git protocol, but like IMAP, the protocol version is specified as a
+capability.  This document defines protocol version 1, which is specified as
+`version=1`.  If the server supports other protocol versions, it may enumerate
+them here as well.
+
+The client will then issue an appropriate version command:
+
+```
+version-request = PKT-LINE("version " number LF) flush-pkt
+number = 1*DIGIT
+```
+
+No reply is expected.
```

Comment:
I think that further explanation here would be good. What if the server refuses to serve a protocol request? I assume that it would terminate the connection, but perhaps there is a convention for agreeing upon a greater-common-denominator-esque protocol(s)?

Reply:
Yeah, I thought about that.  I assumed that the client wouldn't specify a version which wasn't implemented, but we will have to deal with people who intentionally choose to provide invalid data, so we might as well specify something.